# alexrainbirdMusic Spotify Recommender System (Part 1 of 3) 

The goal of this project is to build a recommender system that recommends song similar to songs found on user: alexrainbirdMusic's playlist. 

Knowing what song to incorporate onto alexrainbirdMusic's playlists is extremely important since his profitability is tied to the number of views he has on his playlist videos on YouTube. alexrainbirdMusic has over 1.2M subscribers on YouTube and over 116k followers on Spotify. The more playlists he creates that ties to his goal of bringing the "finest independent pop, folk and rock music" to his followers, the more views and profit he will get. 

Currently, alexrainbirdMusic has artists submit their songs on his website and he reviews the song to determine if it should be added to a playlist. This is a much more manual and tedius process for selecting songs. 

This notebook details the gathering of playlist, artist, and track data  from user: alexrainbirdMusic's playlists. The result is two dataframes with this data - one dataframe is for tracks found on alexrainbirdMusic's playlists and the other dataframe is for tracks not found on alexrainbirdMusic's playlists but along similar genres (alternative, indie, pop, folk). 


## Structure: 
**0. Download & Import Packages**: Download and import relevant packages. 

**1. Use Spotify API to Extract Playlist Data**: To get all alexrainbirdMusic's playlist (300) names, uris, and urls, use cURL and the command line to implement the Spotify Web API code.

**2. Get Track Data by Playlist**: UseSpotipy to get every track and their related details off of each playlist.

**3. Save Data to CSV**: Save the dataframes to csv. 

# Credits
This notebook is built on top of:
- Cameron Watt's and Eric Chang's: https://github.com/enjuichang/PracticalDataScience-ENCA/blob/main/notebooks/Extract%20Features%20Script.ipynb
- Spotify API documentation: https://developer.spotify.com/documentation/web-api/reference/#/operations/get-playlist
- Spotipy documentation: https://spotipy.readthedocs.io/en/2.22.1/

# 0. Download & Import Packages

In [1]:
#!pip install spotipy

In [37]:
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns
import json
import numpy as np
import warnings
warnings.simplefilter("ignore", UserWarning)


import spotipy
import spotipy.util as util
from spotipy.oauth2 import SpotifyClientCredentials
from spotipy.oauth2 import SpotifyOAuth

import datetime
import time

# 1. Use Spotify API to Extract Playlist Data

Using the Spotify API and cURL on the command line, I extracted the playlist names, playlist links, and playlist uris from alexrainbirdMusic's user profile in 50 item increments. This allowed me to quickly retrieve the playlist information to extract track data. 

For the other playlists, I combed Spotify to find relevant playlists to use for the recommendation model. If songs on these playlist overlap with the alexrainbirdMusic playlists, using that similarity the recommendation model will recommend songs related.

In [38]:
# authorize spotipy credentials 
cid = "51eea67316e841838ae183e36fd524f2"
secret = "f990387b1ffd467dad1c4f0737ff88fb"

client_credentials_manager = SpotifyClientCredentials(client_id=cid, client_secret=secret)
sp = spotipy.Spotify(client_credentials_manager = client_credentials_manager)

In [ ]:
#https://maxpiersanti.medium.com/automating-spotify-with-python-everything-you-need-to-know-about-the-spotipy-library-f84d60e8a3ec
import spotipy
import spotipy.util as util
from spotipy.oauth2 import SpotifyClientCredentials
from spotipy.oauth2 import SpotifyOAuth

cid = '5b16d137d0ef4780803e9e43674c0bff'
secret = 'd3815c7bc5f5447ea2c42cf668084251'
scope = "playlist-modify-public"
token = util.prompt_for_user_token('1255636379',scope,client_id=cid,client_secret=secret,redirect_uri='https://example.com/')
sp = spotipy.Spotify(auth=token)


In [ ]:
# utilize Spotify API to retrieve token 
#import requests

#client_id = cid # Enter your client id here
#client_secret = secret# Enter your client secret here

#grant_type = 'client_credentials'

#Request based on Client Credentials Flow from https://developer.spotify.com/web-api/authorization-guide/

#Request body parameter: grant_type Value: Required. Set it to client_credentials
#body_params = {'grant_type' : grant_type}

#url='https://accounts.spotify.com/api/token'

#response=requests.post(url, data=body_params, auth = (client_id, client_secret)) 
#print(response)

#token = json.loads(response.text)['access_token']
#token

In [ ]:
### command line code to get playlist data 

## update the token with new token as needed 

## first json file
#curl -X "GET" "https://api.spotify.com/v1/users/alxrnbrdmusic/playlists?limit=50" -H "Accept: application/json" -H "Content-Type: application/json" -H "Authorization: Bearer BQBlza9lUpF1t7xy4IRRZoDupHQePwZyoWtRi-caCa4xVwD7RZIB0tIKOFD7CPMN61mqU5UeMYQjDH6bjssWmW3EA67pD-xjOQM70ZIRN8uUynxEifUH"

## second json file
#curl -X "GET" "https://api.spotify.com/v1/users/alxrnbrdmusic/playlists?limit=50&offset=50" -H "Accept: application/json" -H "Content-Type: application/json" -H "Authorization: Bearer BQAMjxHKsvFYkDUFdbHpvWUMDgwpQ-gQNVKN989m15ViZdbdO18RWoSFMqkQhjooFXFupnyYUllLToOyLhXvKFgYN9ZOxGKTe0hHPD13CK6iO-SKYMCZAclhimFNuctUP4n6gUAfW5NBnllyCkOKfjgHvRy9LWKJh9d-VDbQLitM-O2rVDM1mIVj5cdqSg"

## third json file
#curl -X "GET" "https://api.spotify.com/v1/users/alxrnbrdmusic/playlists?limit=50&offset=100" -H "Accept: application/json" -H "Content-Type: application/json" -H "Authorization: Bearer BQAMjxHKsvFYkDUFdbHpvWUMDgwpQ-gQNVKN989m15ViZdbdO18RWoSFMqkQhjooFXFupnyYUllLToOyLhXvKFgYN9ZOxGKTe0hHPD13CK6iO-SKYMCZAclhimFNuctUP4n6gUAfW5NBnllyCkOKfjgHvRy9LWKJh9d-VDbQLitM-O2rVDM1mIVj5cdqSg"

## fourth json file
#curl -X "GET" "https://api.spotify.com/v1/users/alxrnbrdmusic/playlists?limit=50&offset=150" -H "Accept: application/json" -H "Content-Type: application/json" -H "Authorization: Bearer BQAMjxHKsvFYkDUFdbHpvWUMDgwpQ-gQNVKN989m15ViZdbdO18RWoSFMqkQhjooFXFupnyYUllLToOyLhXvKFgYN9ZOxGKTe0hHPD13CK6iO-SKYMCZAclhimFNuctUP4n6gUAfW5NBnllyCkOKfjgHvRy9LWKJh9d-VDbQLitM-O2rVDM1mIVj5cdqSg"

## fifth json file
#curl -X "GET" "https://api.spotify.com/v1/users/alxrnbrdmusic/playlists?limit=50&offset=200" -H "Accept: application/json" -H "Content-Type: application/json" -H "Authorization: Bearer BQAMjxHKsvFYkDUFdbHpvWUMDgwpQ-gQNVKN989m15ViZdbdO18RWoSFMqkQhjooFXFupnyYUllLToOyLhXvKFgYN9ZOxGKTe0hHPD13CK6iO-SKYMCZAclhimFNuctUP4n6gUAfW5NBnllyCkOKfjgHvRy9LWKJh9d-VDbQLitM-O2rVDM1mIVj5cdqSg"

## sixth json file
#curl -X "GET" "https://api.spotify.com/v1/users/alxrnbrdmusic/playlists?limit=50&offset=250" -H "Accept: application/json" -H "Content-Type: application/json" -H "Authorization: Bearer BQAMjxHKsvFYkDUFdbHpvWUMDgwpQ-gQNVKN989m15ViZdbdO18RWoSFMqkQhjooFXFupnyYUllLToOyLhXvKFgYN9ZOxGKTe0hHPD13CK6iO-SKYMCZAclhimFNuctUP4n6gUAfW5NBnllyCkOKfjgHvRy9LWKJh9d-VDbQLitM-O2rVDM1mIVj5cdqSg"


In [4]:
# import json files from API outputs
import json
# files are limited to 50 using API 
# https://developer.spotify.com/documentation/web-api/reference/#/operations/get-a-list-of-current-users-playlists
json_files = ['playlistoutput1.json', 
              'playlistoutput2_.json', 
              'playlistoutput3.json', 
              'playlistoutput4.json', 
              'playlistoutput5.json',
              'playlistoutput6.json']  
playlist_names = []
playlist_links = []
playlist_uris =[]

for file in json_files: 
    # open json file
    f = open(file, encoding="utf8")

    # returns JSON object as 
    # a dictionary
    jsonData = json.load(f)
    
    # extract playlist name and url
    for x in range(len(jsonData['items'])):
        name = jsonData["items"][x]["name"]
        playlist_names.append(name)

        url = jsonData["items"][x]["external_urls"]['spotify']
        playlist_links.append(url)
        
        uri = jsonData["items"][x]['uri'][len("spotify:playlist:"):]
        playlist_uris.append(uri)

    print(file + ' data extracted')
print (f"# of playlist names: {len(playlist_names)}")
print (f"# of playlist urls: {len(playlist_links)}")
print (f"# of playlist uris: {len(playlist_uris)}")

print('Successfully extracted urls from playlist outputs json')

playlistoutput1.json data extracted
playlistoutput2_.json data extracted
playlistoutput3.json data extracted
playlistoutput4.json data extracted
playlistoutput5.json data extracted
playlistoutput6.json data extracted
# of playlist names: 300
# of playlist urls: 300
# of playlist uris: 300
Successfully extracted urls from playlist outputs json


In [5]:
len(set(playlist_names))

300

In [6]:
# create playlist df with names, links, and uris (used for track extracting)
arb_dict = {'name': playlist_names, 'links': playlist_links, 'playlist_uris': playlist_uris, 'user': 'alexrainbirdMusic'}
arb = pd.DataFrame(data=arb_dict)

# drop playlists that are podcasts 
arb = arb[arb["name"].str.contains("podcast", case = False) == False]
arb.set_axis(range(len(arb)), inplace=True) #reset index
arb

name  \
0    Indie/Rock/Alt Compilation - February 2023 (al...   
1    Indie/Pop/Folk Compilation - February 2023 (al...   
2    Acoustic Breakfast 🥣 - An Indie/Pop/Folk Playl...   
3                               alexrainbirdMusic 2023   
4    Best of alexrainbirdMusic 2022 🎉 - An Indie/Po...   
..                                                 ...   
294                      Indie/Folk/Alt Halloween 2016   
295         Ultimate Indie-Pop Playlist (2016 Edition)   
296        Ultimate Indie-Folk Playlist (2016 Edition)   
297                             alexrainbirdMusic 2016   
298                           alexrainbirdMixtape 2015   

                                                 links  \
0    https://open.spotify.com/playlist/30uvudTS1E54...   
1    https://open.spotify.com/playlist/3xXory7XEbpS...   
2    https://open.spotify.com/playlist/54Rz7lKoPoVB...   
3    https://open.spotify.com/playlist/51rZ4Mf9G2mw...   
4    https://open.spotify.com/playlist/50xyeCsfk2qj...   
..                                                 ...   
294  https://open.spotify.com/playlist/5BPy1lSZWTot...   
295  https://open.spotify.com/playlist/6HaOOyOEvugF...   
296  https://open.spotify.com/playlist/5fygVerVQfcz...   
297  https://open.spotify.com/playlist/5kiNj8BLGBfK...   
298  https://open.spotify.com/playlist/0IuU2diGoEvO...   

              playlist_uris               user  
0    30uvudTS1E54LdMnpg3ac1  alexrainbirdMusic  
1    3xXory7XEbpSDsA88qeMM7  alexrainbirdMusic  
2    54Rz7lKoPoVBsSxjlHPCmY  alexrainbirdMusic  
3    51rZ4Mf9G2mw6Vaux8sxQb  alexrainbirdMusic  
4    50xyeCsfk2qjMzvx8hrUqu  alexrainbirdMusic  
..                      ...                ...  
294  5BPy1lSZWTotS7mp6cpdxS  alexrainbirdMusic  
295  6HaOOyOEvugF3ZOzHOrE6z  alexrainbirdMusic  
296  5fygVerVQfczi3hrfw5Zkn  alexrainbirdMusic  
297  5kiNj8BLGBfKLwASSxLu2U  alexrainbirdMusic  
298  0IuU2diGoEvOm6zEt8eu3h  alexrainbirdMusic  

[299 rows x 4 columns]

In [7]:
names_notarb = ["New Alt-Rock Mixtape",
       "alt.indie rising",
       "Alternative Rock's Most Precious Gems",
       "All New Indie",
       "Best Rising Alternative Rock",
       "Indie Rising",
       "Roots Rising",
       "Folk Rizing",
       "Indie Pop. Coffee House. Aesthetic. Chillwave. Alternative R&B. Retro",
       "indie pop"]

links_notarb = ["https://open.spotify.com/playlist/37i9dQZF1DXa6YOhGMjjgx?si=7308b7fa62e444a5",
       "https://open.spotify.com/playlist/5NFLDpX0KpIbluIIp37AJy?si=7a45762221534407g",
        "https://open.spotify.com/playlist/37i9dQZF1DXcKHVXMYcCuJ?si=fbd9fe3cc9b94d1e",
        "https://open.spotify.com/playlist/37i9dQZF1DXdbXrPNafg9d?si=c639daef9dfd4528",
        "https://open.spotify.com/playlist/5ZOmKNhdc1Nv94zwScDs59?si=307345dc00314082",
        "https://open.spotify.com/playlist/2Jn2Fv8hWUDchcSyeovjjo?si=f72900661a134aa6",
        "https://open.spotify.com/playlist/37i9dQZF1DWYV7OOaGhoH0?si=49c4f54408cd4007",
        "https://open.spotify.com/playlist/6tkwPMzaA1qzCEwvdnpaJt?si=16eedf7f80304bcb",
        "https://open.spotify.com/playlist/3AKGTEanSdOnQEWOVTMghD?si=a9ac677efdce4b2c",
        "https://open.spotify.com/playlist/6pVBhLF5YLEh7ytCaF3rxo?si=29cbdd7836e549db"]

user_notarb = ["Spotify",
       "Matthewy Keys",
       "Spotify",
       "Spotify",
       "iamlrene",
       "Indie Music Playlists",
       "Spotify",
       "Rizing Playlists",
       "Not Your Bae",
       "Ian"]

# get the playlist IDs 
playlist_uri_notarb = []
for url in links_notarb:     
    link = url.split("/")[-1].split("?")[0]
    playlist_uri_notarb.append(link)


not_arb_dict = {'name': names_notarb, 'links': links_notarb, 'playlist_uris': playlist_uri_notarb, 'user': user_notarb}
not_arb = pd.DataFrame(data=not_arb_dict)
not_arb

name  \
0                               New Alt-Rock Mixtape   
1                                   alt.indie rising   
2              Alternative Rock's Most Precious Gems   
3                                      All New Indie   
4                       Best Rising Alternative Rock   
5                                       Indie Rising   
6                                       Roots Rising   
7                                        Folk Rizing   
8  Indie Pop. Coffee House. Aesthetic. Chillwave....   
9                                          indie pop   

                                               links           playlist_uris  \
0  https://open.spotify.com/playlist/37i9dQZF1DXa...  37i9dQZF1DXa6YOhGMjjgx   
1  https://open.spotify.com/playlist/5NFLDpX0KpIb...  5NFLDpX0KpIbluIIp37AJy   
2  https://open.spotify.com/playlist/37i9dQZF1DXc...  37i9dQZF1DXcKHVXMYcCuJ   
3  https://open.spotify.com/playlist/37i9dQZF1DXd...  37i9dQZF1DXdbXrPNafg9d   
4  https://open.spotify.com/playlist/5ZOmKNhdc1Nv...  5ZOmKNhdc1Nv94zwScDs59   
5  https://open.spotify.com/playlist/2Jn2Fv8hWUDc...  2Jn2Fv8hWUDchcSyeovjjo   
6  https://open.spotify.com/playlist/37i9dQZF1DWY...  37i9dQZF1DWYV7OOaGhoH0   
7  https://open.spotify.com/playlist/6tkwPMzaA1qz...  6tkwPMzaA1qzCEwvdnpaJt   
8  https://open.spotify.com/playlist/3AKGTEanSdOn...  3AKGTEanSdOnQEWOVTMghD   
9  https://open.spotify.com/playlist/6pVBhLF5YLEh...  6pVBhLF5YLEh7ytCaF3rxo   

                    user  
0                Spotify  
1          Matthewy Keys  
2                Spotify  
3                Spotify  
4               iamlrene  
5  Indie Music Playlists  
6                Spotify  
7       Rizing Playlists  
8           Not Your Bae  
9                    Ian

# 2. Get Track Data by Playlist 

Using spotipy, extract track data and audio features from songs on the 299 alexrainbirdMusic playlists and 10 other playlists.

**Key Features Collected:** 
- **artist:** Artist name
- **artist_id:** Artist URI
- **artist_pop**: Artist popularity (0 least popular to 100 most popular)
- **artist_genres**: Genres the artist is associated with
- **album**: Album the track is off of 
- **album_id**: Album URI
- **track_name**: Track name
- **track_id**: Track URI
- **track_pop**: Track popularity (0 least popular to 100 most popular)
- **danceability**: 0 least danceable to 1.0 most danceable
- **energy**: 0 low energy, intensity, and activity to 1.0 most energy, intensity, and activity
- **key**: -1 no pitch detected to 11 highest key
- **loudness**: -60 low decibels to 0 high decibles
- **mode**: 0 minor scale to 1.0 major scale
- **speechiness**: 0 no spoken words to 1.0 exclusive speech-like 
- **acousticness**: 0 no acousticness to 1.0 acoustic
- **instrumentalness**: 0 higher vocals to 10.0 no vocals, only instruments
- **liveness**: 0 no audience in the recording to 1.0 track was performed live
- **valence**: 0 more negative, sad, depressed and/or angry to 1.0 more positive, happy, cheerful, and/or euphoric 
- **tempo**: 0 low beats per minutes to 1.0 higher beats per minute
- **duration_ms**: Length of track in miliseconds
- **time_signature**: Beats per measure

In [26]:
def analyze_playlist(playlist_id):
    
    '''
    Collect artist, album, track, and audio features data from a given playlist.
    ---
    Input:
    playlist_id: Playlist URI
    
    Output: 
    playlist_df: Dataframe that contains the artist, album, track, and audio features data for each track for the playlist.
    
    '''
    
    # Create empty dataframe
    playlist_features_list = ["artist", "artist_id", "artist_pop", "artist_genres", "album", "album_id", "track_name", "track_id", 
                             "track_pop","danceability", "energy", "key", "loudness", "mode", "speechiness", "acousticness",
                             "instrumentalness", "liveness", "valence", "tempo", "duration_ms", "time_signature"]
    playlist_df = pd.DataFrame(columns = playlist_features_list)
    
    # Create empty dict
    playlist_features = {}
    
    # Loop through every track in the playlist, extract features and append the features to the playlist df
    playlist = sp.playlist_tracks(playlist_id)["items"]
    
    for track in playlist:
        # get metadata
        
        ## get artist info
        try:
            playlist_features["artist"] = track["track"]["album"]["artists"][0]["name"]
        except:
            pass
        
        try:
            playlist_features["artist_id"] = track["track"]["album"]["artists"][0]["id"]
        except:
            pass
        
        # https://towardsdatascience.com/extracting-song-data-from-the-spotify-api-using-python-b1e79388d50
        try:
            #artist_uri = playlist_features["artist_id"]
            artist_info = sp.artist(playlist_features["artist_id"])
            playlist_features["artist_pop"] = artist_info['popularity']
        except:
            pass
        
        try:
            playlist_features["artist_genres"] = [artist_info['genres']]
            
            
            #playlist_features["artist_genres"] = artist_info["genres"]

        except:
            pass
        
        ## get album info

        try:
            playlist_features["album"] = track["track"]["album"]["name"]
        except:
            pass
        
        try:
            playlist_features["album_id"] = track["track"]["album"]["id"]
        except:
            pass
        
        ## get track info

        try:
            playlist_features["track_name"] = track["track"]["name"]
        except:
            pass
        
        try:
            playlist_features["track_id"] = track["track"]["id"]
        except:
            pass
        
        try:
            playlist_features["track_pop"] = track["track"]["popularity"]
        except:
            pass
        
        
        # Get audio features
        try:
            audio_features = sp.audio_features(playlist_features["track_id"])[0]
        except:
            pass

        try:
            for feature in playlist_features_list[9:]:
                playlist_features[feature] = audio_features[feature]
        except: 
            pass
        # Concat the dfs
        track_df = pd.DataFrame(playlist_features, index = [0])
        playlist_df = pd.concat([playlist_df, track_df], ignore_index = True)
        #time.sleep(20)
    return playlist_df

# https://towardsdatascience.com/how-to-create-large-music-datasets-using-spotipy-40e7242cc6a6
def analyze_playlist_dict(playlist_dict):
    
    '''
    Loop through the dictionary of playlist to collect playlist features and return a dataframe.
    ---
    Input:
    playlist_dict: Dictionary of playlist names
    
    Output:
    playlist_dict_df: Completed dataframe that contains the artist, album, track, and audio 
    features data for each track of all playlists in the dictionary.
    
    '''
    
    # Loop through every playlist in the dict and analyze it
    for i, (key, val) in enumerate(playlist_dict.items()):
        playlist_df = analyze_playlist(val)
        print(key)
        # Add a playlist column so that we can see which playlist a track belongs too
        playlist_df["playlist"] = key
        # Create or concat df
        if i == 0:
            playlist_dict_df = playlist_df
        else:
            playlist_dict_df = pd.concat([playlist_dict_df, playlist_df], ignore_index = True)
        #time.sleep(10)
    return playlist_dict_df

## 2a. AlexrainbirdMusic Playlist Track Data Collection 

In [ ]:
# create playlist dict for alexrainbird 
names = arb.name.to_list()
playlist_uri = arb.playlist_uris.to_list()
playlist_dict = dict(zip(names,playlist_uri))
playlist_dict

In [ ]:
# collect track data from playlist uris and create a dictionary
%time
track_data = analyze_playlist_dict(playlist_dict)
track_data.head()

In [27]:
# collect track data from playlist uris and create a dictionary
%time
names = arb.name.to_list()[:50]
playlist_uri = arb.playlist_uris.to_list()[:50]
playlist_dict = dict(zip(names,playlist_uri))
track_data_1 = analyze_playlist_dict(playlist_dict)
track_data_1.head()

Wall time: 0 ns
Indie/Rock/Alt Compilation - February 2023 (alexrainbirdMusic)
Indie/Pop/Folk Compilation - February 2023 (alexrainbirdMusic)
Acoustic Breakfast 🥣 - An Indie/Pop/Folk Playlist to start your day
alexrainbirdMusic 2023
Best of alexrainbirdMusic 2022 🎉 - An Indie/Pop/Folk/Rock Playlist
Indie/Indie-Folk Compilation - Winter 2022/2023 ❄️ (alexrainbirdMusic)
The Indie Chart ⭐ OUR TOP 10 SONGS RIGHT NOW ⭐ | alexrainbirdMusic (Pop, Folk, Rock)
Beth's Indie Favourites! 💃 | alexrainbirdMusic
alexrainbirdRecords - The Complete Catalogue 💿
Indie Pop 🎤 |  alexrainbirdMusic
Folk/Acoustic 🪕 |  alexrainbirdMusic
Indie Rock/Alt 🎸 | alexrainbirdMusic
Instrumental Indie-Folk 🪕 - An Acoustic/Chill Playlist for study, relax and focus
Wanderlust - Music for Travelling 🌲 | alexrainbirdMusic (Indie, Folk, Pop)
Mountains and Oceans ⛰️🌊 | alexrainbirdMusic (Indie, Folk, Pop, Acoustic)
Happy Indie 😃 | alexrainbirdMusic (Folk, Pop)
Sad Indie Songs 😢 | alexrainbirdMusic (Folk, Pop, Chill)
night dri

artist               artist_id artist_pop  \
0             SYML  6AyATGg7mDgBlZ4N5uNog0         74   
1  Various Artists  0LyfQWJT6nXafLPZqxe9Of          0   
2        Boo Seeka  1SFz3S9eSUTc49ysstadiO         53   
3     Quiet Houses  6oeIyvCenamQzsTMYnuZTC         31   
4   Juliana Madrid  6RhkgeqhRai3jy4ULSlxFx         31   

                                       artist_genres  \
0                                              [pop]   
1                                                 []   
2  [aussietronica, australian indie, australian pop]   
3                                                 []   
4                                                 []   

                                        album                album_id  \
0                                     Chariot  3R9K985Qq28VqqzeV65bZ0   
1  Indie / Rock / Alt Compilation: March 2023  2jYAyFnU1j39P7dQ5vhE7u   
2                                     Stories  5C6jaEd7bW1pHkpEsXmAO7   
3                              Hot and Clumsy  4mVZsDCMoexn0hrJxFy9f6   
4                                     Madonna  2uNNQws3ix4ZyVxueGd2un   

       track_name                track_id track_pop  danceability  ...  mode  \
0         Chariot  1ZHboJHdP97xyCaQWuP9h6        50         0.516  ...     1   
1      Embrace It  6bGS1VVUZk5MyQ9pYw2a8Q        11         0.572  ...     1   
2         Stories  0CWDijxzzTvMlp4c4U2LIS        46         0.815  ...     0   
3  Hot and Clumsy  50aFetaKwC3pKTLyUWh7UZ        34         0.480  ...     1   
4         Madonna  1eCdsyHZVqHSi5jhwc4TLs        35         0.618  ...     0   

  speechiness  acousticness instrumentalness  liveness  valence    tempo  \
0      0.0545      0.003280           0.0202    0.0937    0.556  168.027   
1      0.0473      0.000924              0.0    0.0504    0.444  128.988   
2      0.0453      0.009620         0.000088    0.1150    0.949  127.981   
3      0.0395      0.007000           0.0133    0.3100    0.782  160.036   
4      0.0347      0.529000              0.0    0.0695    0.475  104.983   

   duration_ms  time_signature  \
0       200693               4   
1       206047               4   
2       164074               4   
3       271600               4   
4       192133               4   

                                            playlist  
0  Indie/Rock/Alt Compilation - February 2023 (al...  
1  Indie/Rock/Alt Compilation - February 2023 (al...  
2  Indie/Rock/Alt Compilation - February 2023 (al...  
3  Indie/Rock/Alt Compilation - February 2023 (al...  
4  Indie/Rock/Alt Compilation - February 2023 (al...  

[5 rows x 23 columns]

In [29]:
track_data_1.to_csv('track_data_1.csv', index=False)

In [31]:
# collect track data from playlist uris and create a dictionary
%time
names = arb.name.to_list()[50:100]
playlist_uri = arb.playlist_uris.to_list()[50:100]
playlist_dict = dict(zip(names,playlist_uri))
track_data_2 = analyze_playlist_dict(playlist_dict)
track_data_2.head()

Wall time: 0 ns
Indie/Pop/Folk Compilation - June 2022 (alexrainbirdMusic)
Under The Stars 🌌 - A Celestial Folk/Pop Playlist
Indie/Rock/Alt Compilation - May 2022 (alexrainbirdMusic)
Indie/Pop/Folk Compilation - May 2022 (alexrainbirdMusic)
Calm Places 🕯️ - A Safe & Comforting Folk/Pop Playlist
Into The Woods 🌲 - A Mysterious Folk/Pop Playlist
Indie/Rock/Alt Compilation - April 2022 (alexrainbirdMusic)
Indie/Pop/Folk Compilation - April 2022 (alexrainbirdMusic)
Indie/Indie-Folk Compilation - Spring 2022 🌼 (alexrainbirdMusic)
Indie/Rock/Alt Compilation - March 2022 (alexrainbirdMusic)
Indie/Pop/Folk Compilation - March 2022 (alexrainbirdMusic)
Read & Relax 📖 - A Peaceful Folk/Pop Playlist
Indie/Rock/Alt Compilation - February 2022 (alexrainbirdMusic)
Indie/Pop/Folk Compilation - February 2022 (alexrainbirdMusic)
Rainy Days 🌧️ - A Melancholic Folk/Pop Playlist
Indie/Rock/Alt Compilation - January 2022 (alexrainbirdMusic)
Indie/Pop/Folk Compilation - January 2022 (alexrainbirdMusic)
Best 

artist               artist_id artist_pop  \
0     Riley Pearce  0A3HlWZGV8WrCcqxKM2neg         52   
1        Companion  5oG2EMrSn6KA6HIwykWfmo         23   
2  Various Artists  0LyfQWJT6nXafLPZqxe9Of          0   
3          Hazlett  1zO3MgzmcwZLLNUQqeU2XH         53   
4    The Dramatics  1zpo5nF4ALcZ22YB8JOigX         12   

                                       artist_genres  \
0  [australian singer-songwriter, indie anthem-fo...   
1                                                 []   
2                                                 []   
3    [brisbane indie, indie anthem-folk, indie folk]   
4                                                 []   

                                    album                album_id  \
0                             Keep Moving  06lFUQTcMPK4YtlU2cdTqb   
1                         In Every Moment  6cMtcmSZd5wLezB0khX6qH   
2  Indie/Pop/Folk Compilation - June 2022  7e7iWRTsHZS8SXQuNPnocS   
3                         Please Don’t Be  2cRNTlXxqKqxO53FBFtto7   
4                              Magic Hour  1nZsPq26fAflbY34sqytUO   

        track_name                track_id track_pop  danceability  ...  mode  \
0      Keep Moving  7MXvN7VcRzylgnmlRDAZqZ        21         0.442  ...     1   
1  In Every Moment  4zIJ1I67yP5RXoELVnZHlb        12         0.442  ...     1   
2     The Long Way  2U97Ir3Hgs500d8p8bJTsO         3         0.408  ...     1   
3  Please Don't Be  1XHejDLvn1Dw878Can4l9P        50         0.577  ...     1   
4       Magic Hour  4KiNxDt8UNJA7HeIh0nBf5        19         0.398  ...     1   

  speechiness  acousticness instrumentalness  liveness  valence    tempo  \
0      0.0602        0.2770            0.137    0.1100    0.325   83.462   
1      0.0287        0.0692              0.0    0.3180    0.278  169.090   
2      0.0342        0.0140              0.0    0.3780    0.566  143.912   
3      0.0357        0.7930          0.00146    0.1080    0.134  129.839   
4      0.0306        0.8830           0.0165    0.0968    0.173  139.847   

   duration_ms  time_signature  \
0       262349               4   
1       157988               3   
2       213579               4   
3       175202               4   
4       215627               4   

                                            playlist  
0  Indie/Pop/Folk Compilation - June 2022 (alexra...  
1  Indie/Pop/Folk Compilation - June 2022 (alexra...  
2  Indie/Pop/Folk Compilation - June 2022 (alexra...  
3  Indie/Pop/Folk Compilation - June 2022 (alexra...  
4  Indie/Pop/Folk Compilation - June 2022 (alexra...  

[5 rows x 23 columns]

In [32]:
track_data_2.to_csv('track_data_2.csv', index=False)

In [35]:
# collect track data from playlist uris and create a dictionary
%time
names = arb.name.to_list()[100:150]
playlist_uri = arb.playlist_uris.to_list()[100:150]
playlist_dict = dict(zip(names,playlist_uri))
track_data_3 = analyze_playlist_dict(playlist_dict)
track_data_3.head()

Wall time: 0 ns
beautiful. 🗻 - An Emotional Indie/Folk/Pop Playlist | Vol. 2
Indie/Rock/Alt Compilation - April 2021 (alexrainbirdMusic)
Indie/Pop/Folk Compilation - April 2021 (alexrainbirdMusic)
Indie/Indie-Folk Compilation - Spring 2021 🌼 (alexrainbirdMusic) 
Ultimate Best of alexrainbirdMusic (1M Subscribers Playlist) 🎉
alexrainbirdMusic Live Celebration 🎤🥳 (1 Million Subscribers Playlist)
Indie/Rock/Alt Compilation - March 2021 (alexrainbirdMusic)
Indie/Pop/Folk Compilation - March 2021 (alexrainbirdMusic)
Tender Love ❤️ - An Indie/Folk/Pop Playlist | Vol. 3
Indie/Rock/Alt Compilation - February 2021 (alexrainbirdMusic)
Indie/Pop/Folk Compilation - February 2021 (alexrainbirdMusic)
Acoustic Office 🪕🖥️ - An Indie/Folk/Pop Working Playlist
Indie/Rock/Alt Compilation - January 2021 (alexrainbirdMusic)
Indie/Pop/Folk Compilation - January 2021 (alexrainbirdMusic)
alexrainbirdMusic 2020
Indie Christmas 2020 🎄 - A Festive Folk/Pop Playlist
Indie/Rock/Alt Compilation - December 2020 (ale

artist               artist_id artist_pop  \
0  Tyler Brown Williams  5xI8nUpNVKFiyrmcq4l9s7         32   
1       The Paper Kites  79hrYiudVcFyyxyJW0ipTy         70   
2              Voyageur  4e96iK3tKVtBuCdQPjoik8         36   
3           Jordan Hart  2NuHJNDBd9Vq4k09qrXibo         49   
4             K E M A L  5Fd7uQWqAWTmcOidyOUhDo         51   

                                       artist_genres  \
0                                    [indiecoustica]   
1  [australian indie folk, indie folk, stomp and ...   
2                                [modern indie folk]   
3                                [modern indie folk]   
4                                      [swedish pop]   

                              album                album_id  \
0                      Never Let Go  0JPqs21tIWq25cOmjdrjlb   
1  Walk Above the City (feat. MARO)  437GKySGOeGJ223PSQLQ6K   
2                               You  4h3UxHwEIcgcSnBweDTz7X   
3          Only Pieces of the Truth  3FJAo9RwlCuoi046GQZIcH   
4                           moments  32Ce7jGpez2g66TtY6rCnd   

                         track_name                track_id track_pop  \
0                      Never Let Go  46UslmWIYHWp5Irp3xfFMD        24   
1  Walk Above the City (feat. MARO)  6qRIH2KDx2NW81J2PR5Lke         0   
2                               You  42D0Oyb2FP2z8Qh6AFvc8U        14   
3          Only Pieces of the Truth  3fbtZVi3mfFASjkOKKJRDs         9   
4                           moments  58lu3AHXTB5KEzwDaoBF4J        27   

   danceability  ...  mode speechiness  acousticness instrumentalness  \
0         0.282  ...     1      0.0317         0.568          0.00113   
1         0.685  ...     1      0.0272         0.870            0.485   
2         0.363  ...     1      0.0306         0.690            0.179   
3         0.486  ...     1      0.0391         0.728              0.0   
4         0.417  ...     1      0.0389         0.920         0.000078   

   liveness  valence    tempo  duration_ms  time_signature  \
0    0.0885   0.1230  147.953       220373               3   
1    0.1090   0.2920   99.016       240171               4   
2    0.0925   0.0843  143.923       294240               4   
3    0.0909   0.2740  126.904       262959               4   
4    0.0982   0.1170   60.622       191093               4   

                                            playlist  
0  beautiful. 🗻 - An Emotional Indie/Folk/Pop Pla...  
1  beautiful. 🗻 - An Emotional Indie/Folk/Pop Pla...  
2  beautiful. 🗻 - An Emotional Indie/Folk/Pop Pla...  
3  beautiful. 🗻 - An Emotional Indie/Folk/Pop Pla...  
4  beautiful. 🗻 - An Emotional Indie/Folk/Pop Pla...  

[5 rows x 23 columns]

In [36]:
track_data_3.to_csv('track_data_3.csv', index=False)

In [39]:
# collect track data from playlist uris and create a dictionary
%time
names = arb.name.to_list()[150:200]
playlist_uri = arb.playlist_uris.to_list()[150:200]
playlist_dict = dict(zip(names,playlist_uri))
track_data_4 = analyze_playlist_dict(playlist_dict)
track_data_4.head()

Wall time: 0 ns
Indie/Rock/Alt Compilation - March 2020 (alexrainbirdMusic)
Indie/Pop/Folk Compilation - March 2020 (alexrainbirdMusic)
Relaxing Sunday Mornings ☕ - An Indie/Folk/Pop Playlist | Vol. 4
Indie/Rock/Alt Compilation - February 2020 (alexrainbirdMusic)
Indie/Pop/Folk Compilation - February 2020 (alexrainbirdMusic)
Goodnight World 🌌 - An Indie/Folk/Chill Sleeping Playlist
Indie/Rock/Alt Compilation - January 2020 (alexrainbirdMusic)
Indie/Pop/Folk Compilation - January 2020 (alexrainbirdMusic)
Relax/Chill Indie ☕ | alexrainbirdMusic (Acoustic, Folk, Pop)
beautiful indie songs 🗻 | alexrainbirdMusic (Folk, Pop, Chill, Acoustic)
Indie Road Trip 🚐 | alexrainbirdMusic
Dark Indie 👻 | alexrainbirdMusic
Indie Coffeeshop ☕ | alexrainbirdMusic
Indie Love ❤️ | alexrainbirdMusic
Indie Studying 📚 | alexrainbirdMusic (Pop, Folk, Acoustic)
Spring Indie 🌼 | alexrainbirdMusic (Folk, Pop, Acoustic)
Summer Indie ☀️ | alexrainbirdMusic
Indie Christmas 2019 🎄 - A Festive Folk/Pop Playlist
Indie/R

artist               artist_id artist_pop  \
0         Billy Otto  68K7z2GTeiBnNteSfDGOzH         32   
1    Various Artists  0LyfQWJT6nXafLPZqxe9Of          0   
2      The Griswolds  3nUbzTTtP2w4HqvT5y5vkV         48   
3    Various Artists  0LyfQWJT6nXafLPZqxe9Of          0   
4  Dancing On Tables  0onRfrkoTz3yrMeP2Ymhhc         21   

                                 artist_genres  \
0                                     [louvor]   
1                                           []   
2  [indie poptimism, modern rock, shimmer pop]   
3                                           []   
4                        [scottish indie rock]   

                                        album                album_id  \
0              Can't Take the Ocean out of Me  3RfYyvxRqoylNhfsTIIB7h   
1  Indie / Rock/ Alt Compilation (March 2020)  4RygALzAqsPHoIbX2HSAHo   
2                              ALL MY FRIENDS  0wJQQBTnYW2zTo5UBuvLof   
3  Indie / Rock/ Alt Compilation (March 2020)  4RygALzAqsPHoIbX2HSAHo   
4                             Not to Need You  0t7bopIsxJdzm0yQj0IWjq   

                       track_name                track_id track_pop  \
0  Can't Take the Ocean out of Me  6oxS5DwSLL4ZnzPADSKmJz         0   
1      Home Is Where the Heart Is  5RGSxEKE4vzRhWBEypPVvU         0   
2                  ALL MY FRIENDS  0xnRyFJFnljL7d69930MCO         0   
3                  Talk Talk Talk  3by7QSrwJPOnnJJMqRXprz         0   
4                 Not to Need You  1HqhXDvZstSm4zqxr6ac5w         7   

   danceability  ...  mode speechiness  acousticness instrumentalness  \
0         0.351  ...     1      0.0487      0.010900                0   
1         0.438  ...     1      0.0654      0.001000         0.000001   
2         0.444  ...     0      0.0916      0.003880                0   
3         0.447  ...     1      0.0444      0.000366          0.00405   
4         0.610  ...     1      0.0261      0.004340          0.00133   

   liveness  valence    tempo  duration_ms  time_signature  \
0    0.4500    0.588  189.662       199000               4   
1    0.1830    0.533  144.059       207563               4   
2    0.0842    0.335  183.883       182796               4   
3    0.0677    0.639  139.987       267936               4   
4    0.1400    0.122   89.994       236520               4   

                                            playlist  
0  Indie/Rock/Alt Compilation - March 2020 (alexr...  
1  Indie/Rock/Alt Compilation - March 2020 (alexr...  
2  Indie/Rock/Alt Compilation - March 2020 (alexr...  
3  Indie/Rock/Alt Compilation - March 2020 (alexr...  
4  Indie/Rock/Alt Compilation - March 2020 (alexr...  

[5 rows x 23 columns]

In [40]:
track_data_4.to_csv('track_data_4.csv', index=False)

In [41]:
# collect track data from playlist uris and create a dictionary
%time
names = arb.name.to_list()[200:250]
playlist_uri = arb.playlist_uris.to_list()[200:250]
playlist_dict = dict(zip(names,playlist_uri))
track_data_5 = analyze_playlist_dict(playlist_dict)
track_data_5.head()

Wall time: 0 ns
Indie/Rock/Alt Compilation - April 2019 (alexrainbirdMusic)
Indie/Pop/Folk Compilation - April 2019 (alexrainbirdMusic)
Best of alexrainbirdMusic // Vol. 4 (600k Subscribers Playlist) 🎉
Indie/Indie-Folk Compilation - Spring 2019 (alexrainbirdMusic)
Indie/Rock/Alt Compilation - March 2019 (alexrainbirdMusic)
Indie/Pop/Folk Compilation - March 2019 (alexrainbirdMusic)
Forests & Firewood 🔥 - An Indie/Folk/Pop Campfire Playlist 🏕️
Indie/Rock/Alt Compilation - February 2019 (alexrainbirdMusic)
Indie/Pop/Folk Compilation - February 2019 (alexrainbirdMusic)
Relaxing Sunday Mornings ☕ - An Indie/Folk/Pop Playlist | Vol. 3
Indie/Rock/Alt Compilation - January 2019 (alexrainbirdMusic)
Indie/Pop/Folk Compilation - January 2019 (alexrainbirdMusic)
alexrainbirdMusic 2019
Indie Christmas 2018 🎄 - A Festive Folk/Pop Playlist
Indie/Rock/Alt Compilation - December 2018 (alexrainbirdMusic)
Indie/Pop/Folk Compilation - December 2018 (alexrainbirdMusic)
Indie/Indie-Folk Compilation - Winte

artist               artist_id artist_pop  \
0     Various Artists  0LyfQWJT6nXafLPZqxe9Of          0   
1        Yukon Blonde  3CdvcTOH01EXzXu96afkSN         35   
2     Various Artists  0LyfQWJT6nXafLPZqxe9Of          0   
3  Future Generations  3wKj5PmSpnrtz9n9hG2QCA         35   
4        Saint Blonde  0nLQlvzkknOFtgjAzBCcqq         33   

                                       artist_genres  \
0                                                 []   
1  [canadian indie, kelowna bc indie, vancouver i...   
2                                                 []   
3                                  [indie poptimism]   
4                                  [denton tx indie]   

                                         album                album_id  \
0  Indie / Rock / Alt Compilation (April 2019)  15ZUFKeBVJKpUQG0T0y3aJ   
1                                 Critical Hit  4OBgCaJqHFS815Gf8a50Hf   
2  Indie / Rock / Alt Compilation (April 2019)  15ZUFKeBVJKpUQG0T0y3aJ   
3                                    Landscape  7I1VWDD8WupY4rhBF2lBEm   
4                                          Shy  1Bs0q9EXeS3axohQoS2Jyq   

             track_name                track_id track_pop  danceability  ...  \
0          Fool for You  2rEqH7SRyZzIXeff2bzQZl         0         0.615  ...   
1  Love The Way You Are  1OCkShZnGpOQ8MT5LVrE9V         0         0.594  ...   
2              The Move  6VSysrfPx5f2thFRKPXzMy         0         0.722  ...   
3             Landscape  7ylzSN6xz3LaHiCeJHVF9y        22         0.624  ...   
4                   Shy  7vgGPP6qNxBXiRd204ywUQ        39         0.616  ...   

   mode speechiness  acousticness instrumentalness  liveness  valence  \
0     1      0.0268       0.10200           0.0347     0.122    0.332   
1     1      0.0396       0.06510         0.000029     0.280    0.892   
2     1      0.0995       0.39700          0.00114     0.324    0.690   
3     1      0.0289       0.00137            0.836     0.398    0.773   
4     0      0.0385       0.01670              0.0     0.179    0.869   

     tempo  duration_ms  time_signature  \
0  106.004       213408               4   
1  184.053       211960               4   
2  107.075       212063               4   
3  104.002       198867               4   
4  108.060       193264               4   

                                            playlist  
0  Indie/Rock/Alt Compilation - April 2019 (alexr...  
1  Indie/Rock/Alt Compilation - April 2019 (alexr...  
2  Indie/Rock/Alt Compilation - April 2019 (alexr...  
3  Indie/Rock/Alt Compilation - April 2019 (alexr...  
4  Indie/Rock/Alt Compilation - April 2019 (alexr...  

[5 rows x 23 columns]

In [42]:
track_data_5.to_csv('track_data_5.csv', index=False)

In [43]:
# collect track data from playlist uris and create a dictionary
%time
names = arb.name.to_list()[250:]
playlist_uri = arb.playlist_uris.to_list()[250:]
playlist_dict = dict(zip(names,playlist_uri))
track_data_6 = analyze_playlist_dict(playlist_dict)
track_data_6.head()

Wall time: 0 ns
Indie/Pop/Folk Compilation - February 2018 (alexrainbirdMusic)
Relaxing Sunday Mornings ☕ - An Indie/Folk/Pop Playlist | Vol. 2
Indie/Rock/Alt Compilation - January 2018 (alexrainbirdMusic)
alexrainbirdMusic 2018
Indie/Pop/Folk Compilation - January 2018 (alexrainbirdMusic)
Indie Christmas 2017 🎄 - A Festive Folk/Pop Playlist
Indie/Rock/Alt Compilation - December 2017 (alexrainbirdMusic)
Indie/Pop/Folk Compilation - December 2017 (alexrainbirdMusic)
Indie/Indie-Folk Compilation - Winter 2017 (alexrainbirdMusic)
Indie/Rock/Alt Compilation - November 2017 (alexrainbirdMusic)
Indie/Pop/Folk Compilation - November 2017 (alexrainbirdMusic)
Dark Forest 🦇 - An Indie/Folk/Alternative Playlist
Indie/Rock/Alt Compilation - October 2017 (alexrainbirdMusic)
Indie/Pop/Folk Compilation - October 2017 (alexrainbirdMusic)
Best of alexrainbirdMusic // Vol. 1 (300k Subscribers Playlist)
Indie/Indie-Folk Compilation - Autumn/Fall 2017 (alexrainbirdMusic)
Indie/Rock/Alt Compilation - Septe

artist               artist_id artist_pop  \
0                Tim Halperin  3nVeK8gQChVtVCCjGreal2         52   
1                 Dusty Boots  4f7aac6rSCC2VopLH049zY         32   
2  Josiah and the Bonnevilles  3FMcVBx2TMq2f5gEPcUieC         60   
3                     Ade Mai  073s2zTquoStQeb482SxhP         20   
4             Daniel Docherty  186gOhAX0pNXjYldk47D0F         37   

                                       artist_genres                 album  \
0  [acoustic pop, indiecoustica, neo mellow, vira...  Forever Starts Today   
1                            [australian indie folk]              Hideaway   
2   [indie folk, modern folk rock, stomp and holler]                 Swing   
3                                [modern indie folk]           Not The One   
4                                [indie anthem-folk]               Hold Me   

                 album_id            track_name                track_id  \
0  7IMJWRBQc7V32AFsr1y2Iu  Forever Starts Today  28cPpB9DbXetlrMkzGGD5v   
1  0gsqzHNTj4LT4zKAkGhCiH              Hideaway  06EJ9feCnHkFc8tSezfyXC   
2  1nB3vRrIRhzVkcQozGEPVQ                 Swing  3EiZbFxFsv0l8WyEmdurqz   
3  4OscTSUTEc1o1vJpfsyMmi           Not The One  0PQFkHo7Np34jUHpxtuwCV   
4  2xndSAjrx6eWmKO8siRUDf               Hold Me  4jQGKnI5CcqZ8POmE88CHF   

  track_pop  danceability  ...  mode speechiness  acousticness  \
0         0         0.463  ...     0      0.0378       0.36000   
1        22         0.663  ...     1      0.0293       0.32900   
2        11         0.456  ...     1      0.0354       0.00207   
3         0         0.606  ...     1      0.0299       0.67200   
4         0         0.385  ...     1      0.0295       0.58000   

  instrumentalness  liveness  valence    tempo  duration_ms  time_signature  \
0          0.00563    0.1970   0.0424   85.035       194556               4   
1          0.00253    0.1240   0.3150  114.968       257021               4   
2              0.0    0.0875   0.5500  168.129       183867               4   
3            0.306    0.1030   0.2300  119.963       241500               4   
4          0.00239    0.0673   0.5260   82.869       226743               4   

                                            playlist  
0  Indie/Pop/Folk Compilation - February 2018 (al...  
1  Indie/Pop/Folk Compilation - February 2018 (al...  
2  Indie/Pop/Folk Compilation - February 2018 (al...  
3  Indie/Pop/Folk Compilation - February 2018 (al...  
4  Indie/Pop/Folk Compilation - February 2018 (al...  

[5 rows x 23 columns]

In [44]:
track_data_6.to_csv('track_data_6.csv', index=False)

In [45]:
track_data_arb = pd.concat([track_data_1, 
                            track_data_2, 
                            track_data_3,
                            track_data_4,
                            track_data_5,
                            track_data_6])

## 2b. Other Playlist Track Data Collection 

In [9]:
# create playlist dict for not alexrainbird 

names = not_arb.name.to_list()
playlist_uri_not_arb = not_arb.playlist_uris.to_list()
playlist_dict_not_arb = dict(zip(names,playlist_uri_not_arb))
playlist_dict_not_arb

{'New Alt-Rock Mixtape': '37i9dQZF1DXa6YOhGMjjgx',
 'alt.indie rising': '5NFLDpX0KpIbluIIp37AJy',
 "Alternative Rock's Most Precious Gems": '37i9dQZF1DXcKHVXMYcCuJ',
 'All New Indie': '37i9dQZF1DXdbXrPNafg9d',
 'Best Rising Alternative Rock': '5ZOmKNhdc1Nv94zwScDs59',
 'Indie Rising': '2Jn2Fv8hWUDchcSyeovjjo',
 'Roots Rising': '37i9dQZF1DWYV7OOaGhoH0',
 'Folk Rizing': '6tkwPMzaA1qzCEwvdnpaJt',
 'Indie Pop. Coffee House. Aesthetic. Chillwave. Alternative R&B. Retro': '3AKGTEanSdOnQEWOVTMghD',
 'indie pop': '6pVBhLF5YLEh7ytCaF3rxo'}

In [10]:
# collect track data from playlist uris and create a dictionary

%time
track_data_not_arb = analyze_playlist_dict(playlist_dict_not_arb)

Wall time: 0 ns
New Alt-Rock Mixtape
alt.indie rising
Alternative Rock's Most Precious Gems
All New Indie
Best Rising Alternative Rock
Indie Rising
Roots Rising
Folk Rizing
Indie Pop. Coffee House. Aesthetic. Chillwave. Alternative R&B. Retro
indie pop


In [11]:
track_data_not_arb.head()

artist               artist_id artist_pop  \
0             Tigercub  6ekYAO2D1JkI58CF4uRRqw         47   
1             Crawlers  2xtmoxSauQs0TQFUoHmbfy         52   
2       The Luka State  6DaXEbr3LdLNcui8pZf6AF         44   
3  The Backseat Lovers  6p2HnfM955TI1bX34dkLnI         70   
4           Fleshwater  6P5ccCJCe8A4s9tDSTNFzF         52   

                                       artist_genres  \
0  [brighton indie, modern alternative rock, mode...   
1                      [indie pop, warrington indie]   
2  [english indie rock, modern alternative rock, ...   
3                [indie pop, modern rock, slc indie]   
4                               [dreamo, grungegaze]   

                        album                album_id              track_name  \
0      Play My Favourite Song  48kUWLzmZrvwHqYrpOkBat  Play My Favourite Song   
1               Crawlers - EP  4wePwIhGnXxJ3tRFAusMAE       Come Over (Again)   
2              More Than This  4xENjLbcy7IdEHA3JQzYRx          More Than This   
3        When We Were Friends  3TSMSh5dai7WEnEGOoMXBZ              Kilby Girl   
4  We're Not Here to Be Loved  0hm7PiBu72tRliLqLfiKy1         Kiss the Ladder   

                 track_id track_pop  danceability  ...  mode speechiness  \
0  4uBs8miGwBykRYjrbAO5kV        42         0.509  ...     1      0.0579   
1  4PDJDIdWxNN1AlnbrKkoPf        60         0.498  ...     1      0.0278   
2  1F3VhVtaMqUqKhXdpA3itF        36         0.443  ...     0      0.1300   
3  1170VohRSx6GwE6QDCHPPH        73         0.329  ...     1      0.0417   
4  41QBT1Al5RQ9u9UIHOuXnj        56         0.122  ...     1      0.1350   

   acousticness instrumentalness  liveness  valence    tempo  duration_ms  \
0      0.001970          0.00633    0.2650    0.523   95.961       155603   
1      0.003930              0.0    0.0675    0.235   93.018       254419   
2      0.000026          0.00332    0.2980    0.333  112.307       182100   
3      0.057800           0.0352    0.1130    0.225  162.279       282206   
4      0.000009         0.000324    0.4030    0.250  174.248        77467   

   time_signature              playlist  
0               4  New Alt-Rock Mixtape  
1               4  New Alt-Rock Mixtape  
2               4  New Alt-Rock Mixtape  
3               4  New Alt-Rock Mixtape  
4               4  New Alt-Rock Mixtape  

[5 rows x 23 columns]

# 3. Save Data

In [12]:
# save data 

# playlist details
arb.to_csv('arb_playlist_details.csv', index = False)
not_arb.to_csv('not_arb_playlist_details.csv', index = False)

In [46]:
# completed track data 
track_data_arb.to_csv('arb_music_final.csv', index=False)
track_data_not_arb.to_csv('not_arb_music_final.csv', index=False)

# END